File: src/query-tree/query-attached.ts
import { Query } from './query'
import { QueryTreeNode } from './query-tree-node'

// AttachedQuery tracks internal information about a query attached to a tree.
export class AttachedQuery {
  // qtNodes are the query tree nodes referenced by this query.
  public qtNodes: QueryTreeNode[] = []

  constructor(
    // query is the underlying query
    public query: Query
  ) {}

  // appendQueryNode appends a query tree node to the attached query.
  public appendQueryNode(qtNode: QueryTreeNode) {
    this.qtNodes.push(qtNode)
  }
}

File: client/query-attached.go
package client

// qtAttachedQuery is a query attached to the query tree.
type qtAttachedQuery struct {
	// query is the referenced query
	query *Query
	// qtNodes are the query tree nodes referenced by this query.
	qtNodes []*qtNode
}

func newQtAttachedQuery(query *Query) *qtAttachedQuery {
	return &qtAttachedQuery{query: query}
}

// appendChildNode adds a qtNode to the query.
func (q *qtAttachedQuery) appendQueryNode(n *qtNode) {
	q.qtNodes = append(q.qtNodes, n)
}
